juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1561 @@
+    Null,
+    Scalar(&'a S),
+    Enum(&'a str),
+    List(LookAheadList<'a, S>),
+    Object(LookAheadObject<'a, S>),
+}
+
+impl<'a, S: ScalarValue + 'a> LookAheadValue<'a, S> {
+    fn from_input_value(
+        input_value: &'a InputValue<S>,
+        span: &'a Span,
+        vars: Option<&'a Variables<S>>,
+    ) -> BorrowedSpanning<'a, Self> {
+        Spanning {
+            item: match input_value {
+                InputValue::Null => Self::Null,
+                InputValue::Scalar(s) => Self::Scalar(s),
+                InputValue::Enum(e) => Self::Enum(e),
+                InputValue::Variable(name) => vars
+                    .and_then(|vars| vars.get(name))
+                    .map(|item| Self::from_input_value(item, span, vars).item)
+                    .unwrap_or(Self::Null),
+                InputValue::List(input_list) => Self::List(LookAheadList { input_list, vars }),
+                InputValue::Object(input_object) => Self::Object(LookAheadObject {
+                    input_object: input_object.as_slice(),
+                    vars,
+                }),
+            },
+            span,
+        }
+    }
+}
+
+/// A JSON-like list that can be used as an argument in the query execution.
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadList<'a, S> {
+    input_list: &'a [Spanning<InputValue<S>>],
+    vars: Option<&'a Variables<S>>,
+}
+
+impl<'a, S: ScalarValue> LookAheadList<'a, S> {
+    /// Returns an iterator over the list's elements.
+    pub fn iter(&self) -> impl Iterator<Item = BorrowedSpanning<'a, LookAheadValue<'a, S>>> {
+        let vars = self.vars;
+        self.input_list
+            .iter()
+            .map(move |val| LookAheadValue::from_input_value(&val.item, &val.span, vars))
+    }
+}
+
+impl<'a, S> Default for LookAheadList<'a, S> {
+    fn default() -> Self {
+        Self {
+            input_list: &[],
+            vars: None,
+        }
+    }
+}
+
+impl<'a, S: ScalarValue> PartialEq for LookAheadList<'a, S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iter().eq(other.iter())
+    }
+}
+
+/// A JSON-like object that can be used as an argument in the query execution.
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadObject<'a, S> {
+    input_object: &'a [(Spanning<String>, Spanning<InputValue<S>>)],
+    vars: Option<&'a Variables<S>>,
+}
+
+impl<'a, S: ScalarValue + 'a> LookAheadObject<'a, S> {
+    /// Returns an iterator over the object's entries.
+    pub fn iter(
+        &self,
+    ) -> impl Iterator<
+        Item = (
+            BorrowedSpanning<'a, &'a str>,
+            BorrowedSpanning<'a, LookAheadValue<'a, S>>,
+        ),
+    > {
+        let vars = self.vars;
+        self.input_object.iter().map(move |(key, val)| {
+            (
+                Spanning {
+                    span: &key.span,
+                    item: key.item.as_str(),
+                },
+                LookAheadValue::from_input_value(&val.item, &val.span, vars),
```

Comment:
@audunhalland so, this is the place where the actual laziness happens. Instead of traversing and converting the whole AST on the root value, we do it in the actual iteration.
However, it seems that it happens each time we do this iteration. So, if we repeatedly work in the same method on the same AST node multiple times, we do parse/convert AST nodes multiple times too.

I wonder whether it can introduce any performance issues to the level we should implement memoization out-of-the-box?

Reply:
I interpret this mapping as very cheap. There's no allocation involved; it involves an enum `match` and the most expensive thing that can happen is one HashMap lookup in case you're looking at a Variable. The rest is just plain reference copying.

It's a completely shallow mapping. Sub-iteration of the contents of lists and objects is something the API user opts in to.

Reply:
@audunhalland thinking about it a little bit more, I agree. Maybe we should just leave a notice in documentation about this, so the library user can memoize the result into a temporary binding, if he does a lot of repeated stuff and needs to.

Reply:
Note about this added in docs for the two iterator types in https://github.com/graphql-rust/juniper/pull/1212/commits/c100140091fb8dbf33ec3396ef4fd8b7bc129d4e.